apis/v1alpha1/backendpolicy_types.go

```diff
@@ -0,0 +1,198 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// BackendPolicy defines policies associated with backends. For the purpose of
+// this API, a backend is defined as any resource that a route can forward
+// traffic to. A common example of a backend is a Service.
+type BackendPolicy struct {
+	metav1.TypeMeta   `json:",inline" protobuf:"bytes,1,opt,name=typeMeta"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,2,opt,name=metadata"`
+
+	Spec   BackendPolicySpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
+	Status BackendPolicyStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+}
+
+// +kubebuilder:object:root=true
+
+// BackendPolicyList contains a list of BackendPolicy
+type BackendPolicyList struct {
+	metav1.TypeMeta `json:",inline" protobuf:"bytes,1,opt,name=typeMeta"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,2,opt,name=metadata"`
+	Items           []BackendPolicy `json:"items" protobuf:"bytes,3,rep,name=items"`
+}
+
+// BackendPolicySpec defines desired policy for a backend.
+type BackendPolicySpec struct {
+	// BackendRefs define which backends this policy should be applied to. If
+	// more than one BackendPolicy targets the same backend, precedence will be
+	// given to the oldest BackendPolicy.
+	//
+	// Support: Core
+	//
+	// +optional
+	BackendRefs []BackendRef `json:"backendRefs" protobuf:"bytes,1,opt,name=backendRefs"`
+
+	// TLS is the TLS configuration for these backends.
+	//
+	// Support: Core
+	//
+	// +optional
+	TLS *TLSConfig `json:"tls,omitempty" protobuf:"bytes,4,opt,name=tls"`
```

Comment:
TLSConfig struct is currently being used for downstream TLS settings for the Gateway.

Using the same struct for Upstream can be confusing, such as, what does CertificateRef mean in this context? Is it the client certificate-key pair that the Gateway should use to authenticate with the upstream Service or is it the CACertificate to use to verify upstream's TLS certificate?

We should ask what settings of upstream TLS we want to make configurable and then create a new struct (and probably rename the (`TLSConfig` struct).

Another question: How/Where is the protocol the Gateway should use to talk to upstream defined?

Reply:
Thanks for catching this, it really didn't make much sense to share the type here. I've split this out into a new BackendTLSConfig that supports references for client cert, key, and trusted CA. Those seemed to be fairly commonly supported but interested in feedback here and/or other fields that should be included.